Consolidate docker and docker-compose

shahbazn opened the pull request:
## Description
Cleaning up the repo to consolidate docker packaging and docker-compose dev setup workflows and remove docker files un related to this repo

## Issues This PR Fixes
Fixes #1960 
Fixes partially #1895 

## How to QA
Follow the updated Contributing.md in this PR

sbellem commented on docker-compose.yml:
```diff
@@ -1,33 +1,45 @@
-      - ./pytest.ini:/usr/src/app/pytest.ini
-      - ./tox.ini:/usr/src/app/tox.ini
     environment:
-      BIGCHAINDB_DATABASE_BACKEND: localmongodb
```
Why is this removed?

shahbazn replied:
we are already passing that as an arg no need for env variable too

ttmc commented:
In `CONTRIBUTING.md` the "Step 2" section needs to be revised as well, since it's currently aimed at getting users set up to run everything as processes, rather than using Docker and Docker Compose, i.e.

<hr>

### Step 2 - Install some Dependencies

Install MongoDB, Tendermint, and all of BigchainDB Server's dependencies. The Quickstart page has some pointers. In fact, you could do everything in the Quickstart page short of installing BigchainDB with pip (since you will install from the source on GitHub), and you shouldn't run MongoDB or Tendermint yet.

<hr>

The only dependencies they'll have now will be Docker and Docker Compose, unless I'm mistaken. You could move the links about how to install Docker and Docker Compose up into the "Step 2" section (out of "Step 6").

<hr>

For people who _really_ want to run everything as processes, maybe you could point them to the `pkg/scripts/` directory?

shahbazn commented on CONTRIBUTING.md:
````diff
@@ -60,29 +60,27 @@ git fetch upstream
+docker-compose -f docker-compose.yml build
+```
+
+NOTE: If you have already ran the above command once you can just run `docker-compose -f docker-compose.yml build` to speed up the build process for any subsequent builds
````
it was supposed to be `docker-compose -f docker-compose.yml build bdb` instead of `docker-compose -f docker-compose.yml build` in the Note section will update. `docker-compose -f docker-compose.yml build` will build the base image bdb-base and also bdb but `docker-compose -f docker-compose.yml build bdb` will only build bdb image without re-building the bdb-base, which is basically just installing bigchaindb code instead of re-installing all the dependencies

shahbazn commented:
@sbellem 
> Role of Dockerfile versus Dockerfile-dev
Added doc strings
>Removal of rdb
Are we still supporting rethinkdb?
>Purpose of bdb-base
Added doc string

@ttmc Addressed most of your comments and for multinode setup with `docker-compose.network.yml` and deployment as process using pkg/scripts I can create seperate PRs for that keep things incremental and manageable?

sbellem commented on Dockerfile-dev:
```diff
@@ -1,24 +1,15 @@
-RUN pip install --no-cache-dir -e .[dev]
-RUN bigchaindb -y configure "$backend"
+
+RUN pip install --no-cache-dir .
```
Why was  `-e .[dev]` removed?

sbellem replied:
The `-e` stands for `editable`. It is similar to `python setup.py develop`:

```bash
$ pip install --help | grep '\-\-editable'
  -e, --editable <path/url>   Install a project in editable mode (i.e. setuptools "develop mode") from a local project path 
```

We need when developing as it allows us to do away with the need of re-installing the project when files are changed.

Thus, the editable mode combined with the mounted volumes in docker-compose allow us to make changes and test them very quickly, without the need to execute `pip install` each time.

sbellem commented on docker-compose.yml:
```diff
@@ -1,33 +1,55 @@
-      - ./setup.py:/usr/src/app/setup.py
-      - ./setup.cfg:/usr/src/app/setup.cfg
-      - ./pytest.ini:/usr/src/app/pytest.ini
-      - ./tox.ini:/usr/src/app/tox.ini
```
Why are these removed?

For instance, `pytest.ini` and `setup.cfg` control `pytest` and are needed when developing and testing.

`docs` are an essential part of development and also need to be mounted.

`setup.py` may be needed to install a new package, and thus it can be useful to mount it.

muawiakh replied:
@shahbazn  can you comment on this?

shahbazn replied:
was just using `docker-compose.tendmint.yml` as the source of truth and replacing the docker-compose.yml with `docker-compose.tendermint.yml` and thats what is reflected here. Now that `docker-compose.tendmint.yml` have been updated will update this

sbellem commented on docker-compose.network.yml:
```diff
@@ -164,7 +164,3 @@ services:
-  driver:
-    build:
-      context: .
-      dockerfile: ./compose/bigchaindb-driver/Dockerfile
```
Why removing the python client service now? This was put there for development purposes and may still be useful as we are not complete with the integration work phase.

ttmc commented on CONTRIBUTING.md:
```diff
@@ -34,7 +34,7 @@ Familiarize yourself with how we do coding and documentation in the BigchainDB p
 ### Step 2 - Install some Dependencies
 
-Install MongoDB, Tendermint, and all of BigchainDB Server's dependencies. The [Quickstart page](https://docs.bigchaindb.com/projects/server/en/latest/quickstart.html) has some pointers. In fact, you could do everything in the Quickstart page short of installing BigchainDB with pip (since you will install from the source on GitHub), and you shouldn't run MongoDB or Tendermint yet.
+We recommend using `Docker` and `Docker Compose` based workflow for testing and development of bigchaindb. So the only dependencies you need to install are [Docker](https://docs.docker.com/engine/installation/) and [docker-compose](https://docs.docker.com/compose/install/) and [Docker Compose](https://docs.docker.com/compose/install/).
```
The second sentence says "...to install are Docker and docker-compose and Docker Compose."

muawiakh replied:
+1, needs to be addressed.

sbellem commented:
@shahbazn and I discussed, and we agreed to put this put PR on hold for now and we'll revisit it after the holidays.

sbellem commented on CONTRIBUTING.md:
````diff
@@ -60,29 +60,27 @@ git fetch upstream
+docker-compose -f docker-compose.yml build
+```
+
+NOTE: If you already ran the above command once you can just run `docker-compose -f docker-compose.yml build bdb` to speed up the build process for any subsequent builds
````
Since the default compose file is `docker-compose.yml` we can avoid the `-f docker-compose.yml` option.

ttmc commented:
> I can create seperate PRs for that keep things incremental and manageable?

Sounds good.

ttmc commented:
I wrote this in an email thread, but it should really be shared here as well:

Shahbaz wanted to make using Docker Compose the standard way of doing development on BigchainDB Server. I figured that was fine so long as there are clear instructions for how to do that (including for new developers, i.e. in `CONTRIBUTING.md`).

Maybe we could document how to do development in a few different ways? Maybe:

1. by installing stuff with apt and pip, using virtualenvs, etc. (i.e. like the current `CONTRIBUTING.md`)
1. using Docker Compose
1. using Vagrant?

Maybe even other ways. I like the idea of making Docker Compose the default since it works cross-platform.

ttmc commented:
(I just approved this PR since I don't want my review to be a blocker while I'm away on holidays.)

muawiakh commented on Dockerfile:
```diff
@@ -1,3 +1,9 @@
+# Main Dockerfile used to build production
```
If you are changing this, you might need to check [Dockerhub](https://hub.docker.com/r/bigchaindb/bigchaindb/~/settings/automated-builds), if we need to pass the argument `backend` at build time.

shahbazn commented:
Will be handled in separate PR.
